File: solution/885. Spiral Matrix III/885.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int y, int x) {
        vector<vector<int>> res = {{y, x}};
        int dy = 0, dx = 1;
        for (int n = 0; res.size() < rows * cols; ++n) {
            for (int i = 0; i < n / 2 + 1; ++i) {
                y += dy, x += dx;
                if (0 <= y && y < rows && 0 <= x && x < cols)
                    res.push_back({y, x});
            }
            // x, y = y, -x
            swap(dy, dx);
            dx = -dx;
        }
        return res;
    }
};

File: solution/885. Spiral Matrix III/885.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, y: int, x: int) -> List[List[int]]:
        res = []
        dy, dx = 0, 1
        
        n = 0
        while len(res) < rows * cols:
            for i in range(n // 2 + 1):
                if 0 <= y < rows and 0 <= x < cols:
                    res.append([y, x])
                y, x = y + dy, x + dx
            dy, dx = dx, -dy
            n += 1
            
        return res
    
